### ISSUE TITLE
Some performance issues with audio encoding

### ISSUE THREAD

**Issue by hexisyztem** [stepfun-ai/Step-Audio#105]
[English]
I understand that step-audio encodes the audio byte stream into audio_token, but I have two questions about this step:
1. Performance: audio encode (should be called StepAudioTokenizer in step_audio) is not good enough. It takes 17 seconds to process a 14s audio using onnxruntime-gpu on L20. This is too time-consuming. Is there a way to optimize it? (Currently in progress)
2. Effect: audio encode directly encodes the byte stream into token. Does this step lose too much information? Is it more reasonable to send it to LLM in the form of multimodal embedding?

----------
[中文]
我理解 step-audio 将音频字节流编码成 audio_token，但是关于这一步我有两个问题: 
1. 性能方面： audio encode(在step_audio中应该被称为 StepAudioTokenizer) 性能并不太足够，在 L20 使用 onnxruntime-gpu 对一段 14s 的音频需要 17秒才能够处理，这耗时过多了，是否有方法进行优化？(目前正在进行中)
2. 效果方面： audio encode 将字节流直接encode成 token，这一步是否丢失了过多的信息，以 multimodal embedding 的形式送入到 llm 是否更为合理

**Comment by hexisyztem** [comment]
As shown in the following code, when StepAudioTokenizer runs on a single thread, it takes a very long time to execute, but when the tokenizer calculation is moved to a separate thread, the time taken to execute is significantly reduced.

如下面的代码所示，当 StepAudioTokenizer 在单一线程上运行时，耗时异常的高，但是当 tokenizer 的计算放到单独的线程时，耗时会明显的下降很多。

```
import time
import os
import torchaudio
from tokenizer import StepAudioTokenizer
from concurrent.futures import ThreadPoolExecutor
import time
import threading

step_audio_tokenizer = StepAudioTokenizer("/mnt3/zhoubofan/Step-Audio/Step-Audio-Tokenizer")

def get_pid_thread_id():
    pid = os.getpid()
    thread_id = threading.current_thread().ident
    return pid, thread_id

def print_all_threads():
    for thread in threading.enumerate():
        print(f"线程名称: {thread.name}, 线程ID: {thread.ident}, 是否存活: {thread.is_alive()}\n\n\n")
    process_id, thread_id = get_pid_thread_id()
    print(f"current_process_id: {process_id}, current_thread_id: {thread_id}")


def step_audio_call():
    audio, sr = torchaudio.load('test_wav.wav') 
    start_time = time.perf_counter()
    speech_tokens_lst = step_audio_tokenizer(audio, sr)
    end_time = time.perf_counter()
    print("-----------> time cost: ", end_time - start_time)
    print_all_threads()

if __name__ == '__main__':
    # while True:
    step_audio_call()

    # while True:
    with ThreadPoolExecutor() as executor:

        future = executor.submit(step_audio_call)
        result = future.result()
        print(result)

```

**Comment by hexisyztem** [comment]
@mabuyun

**Comment by mabuyun** [comment]
@hexisyztem 
I can only answer engineering-related questions such as performance testing. Your testing methodology overlooks the warmup time during the model's first run. Performance testing generally requires running multiple cycles, discarding the timing from the initial run, and then computing the average duration of the subsequent iterations.

我只能回答性能测试等工程上的问题，你的测试方法忽略了模型第一次运行的 warmup 时间，性能测试一般是多次循环后去掉第一次的时间，然后计算之后耗时的平均值

```
loop_cnt = 5
costs = []
audio, sr = torchaudio.load('test_wav.wav')
for _ in range(loop_cnt):
    start_time = time.perf_counter()
    speech_tokens_lst = step_audio_tokenizer(audio, sr)
    end_time = time.perf_counter()
    costs.append(end_time - start_time)
    
average = sum(costs[1:]) / len(costs[1:]) if len(costs[1:]) !=0 else 0

```

**Comment by hexisyztem** [comment]
> [@hexisyztem](https://github.com/hexisyztem) I can only answer engineering-related questions such as performance testing. Your testing methodology overlooks the warmup time during the model's first run. Performance testing generally requires running multiple cycles, discarding the timing from the initial run, and then computing the average duration of the subsequent iterations.
> 
> 我只能回答性能测试等工程上的问题，你的测试方法忽略了模型第一次运行的 warmup 时间，性能测试一般是多次循环后去掉第一次的时间，然后计算之后耗时的平均值
> 
> ```
> loop_cnt = 5
> costs = []
> audio, sr = torchaudio.load('test_wav.wav')
> for _ in range(loop_cnt):
>     start_time = time.perf_counter()
>     speech_tokens_lst = step_audio_tokenizer(audio, sr)
>     end_time = time.perf_counter()
>     costs.append(end_time - start_time)
>     
> average = sum(costs[1:]) / len(costs[1:]) if len(costs[1:]) !=0 else 0
> ```
In fact, the same problem still exists when the above code is changed to multiple calls. I guess it may be a lock competition problem in the tokenizer code logic, but I have not found the specific reason yet.

事实上，上面的代码改为多次调用仍然存在着相同的问题。我猜测可能是 tokenizer 代码逻辑中存在锁竞争的问题，但是目前并没有发现具体原因

**Comment by mabuyun** [comment]
My test results confirmed the warmup issue mentioned earlier, and there was no significant difference in the time consumption between the two invocation methods.

我的测试结果验证了之前提到的warmup问题，两种调用方式的耗时并没有明显差异。


```python
if __name__ == '__main__':
    # while True:
    print("warmup -------------")
    step_audio_call()

    print("1 -------------")
    step_audio_call()

    print("2 -------------")
    # while True:
    with ThreadPoolExecutor() as executor:

        future = executor.submit(step_audio_call)
        result = future.result()
        print(result)

```

```bash
warmup -------------
-----------> time cost:  3.071560438722372
线程名称: MainThread, 线程ID: 139768199919424, 是否存活: True



current_process_id: 56, current_thread_id: 139768199919424
1 -------------
-----------> time cost:  1.0697254501283169
线程名称: MainThread, 线程ID: 139768199919424, 是否存活: True



current_process_id: 56, current_thread_id: 139768199919424
2 -------------
-----------> time cost:  1.0482799410820007
线程名称: MainThread, 线程ID: 139768199919424, 是否存活: True



线程名称: ThreadPoolExecutor-0_0, 线程ID: 139762823898880, 是否存活: True



current_process_id: 56, current_thread_id: 139762823898880
None

```

**Comment by hexisyztem** [comment]
> if __name__ == '__main__':
>     # while True:
>     print("warmup -------------")
>     step_audio_call()
> 
>     print("1 -------------")
>     step_audio_call()
> 
>     print("2 -------------")
>     # while True:
>     with ThreadPoolExecutor() as executor:
> 
>         future = executor.submit(step_audio_call)
>         result = future.result()
>         print(result)

I don't know what's going on.
我也不知道是什么情况了


```
warmup -------------
-----------> time cost:  17.96432103216648
线程名称: MainThread, 线程ID: 140615102895936, 是否存活: True



current_process_id: 347498, current_thread_id: 140615102895936
1 -------------
-----------> time cost:  17.970564179122448
线程名称: MainThread, 线程ID: 140615102895936, 是否存活: True



current_process_id: 347498, current_thread_id: 140615102895936
2 -------------
-----------> time cost:  1.729598343372345
线程名称: MainThread, 线程ID: 140615102895936, 是否存活: True



线程名称: ThreadPoolExecutor-0_0, 线程ID: 140604155876928, 是否存活: True



current_process_id: 347498, current_thread_id: 140604155876928
None
```
